EIPS/eip-4762.md

```diff
@@ -35,155 +35,266 @@ def get_storage_slot_tree_keys(storage_key: int) -> [int, int]:
+      for i in range(nbytes):
+        immediate_chunk = (pc+i) // CHUNK_SIZE
+        if immediate_chunk <= (len(code)+CHUNK_SIZE-1) // CHUNK_SIZE:
+          add_read_access_event(contract_address, (immediate_chunk + CODE_OFFSET) // 256, (immediate_chunk + CODE_OFFSET) % 256)
+
+    case OP_CODECOPY | OP_EXTCODECOPY:
+      addr = contract_address
+      size = 0
+      offset = 0
+      if opcode == OP_EXTCODECOPY:
+        addr = stack.peek(0)
+        offset = stack.peek(2)
+        size = stack.peek(3)
+      else:
+        offset = stack.peek(1)
+        size = stack.peek(2)
+
+      for i in range(size):
+        copied_byte_chunk = (offset+i) // CHUNK_SIZE 
+        if copied_byte_chunk <= (len(code)+CHUNK_SIZE-1) // CHUNK_SIZE:
+          add_read_access_event(contract_address, (copied_byte_chunk + CODE_OFFSET) // 256, (copied_byte_chunk + CODE_OFFSET) % 256)
+
+    case OP_CODESIZE | OP_EXTCODESIZE | OP_EXTCODEHASH:
+      # None of these operations access any chunk, it is
+      # explicitly documented here as a no-op.
+
+  charge_gas()
 ```
-(contract_address, 0, BASIC_DATA_LEAF_KEY)
-(contract_address, 0, CODEHASH_LEAF_KEY)
-```
-
-#### Write events for storage
-
-`SSTORE` opcodes with a given `address` and `key` process a write event of the form
 
-```
-(address, tree_key, sub_key)
-```
+Note: since no access list existed for code up until this EIP, note that no warm costs are charged for code accesses.
 
-Where `tree_key` and `sub_key` are computed as `tree_key, sub_key = get_storage_slot_tree_keys(address, key)`
+When a contract is created, access chunks `0 ... (len(code)+30)//31`:
 
-#### Write events for code
+```python3
+def emit_code_write_access_events(address: Address, code: bytes):
+    # Calculate the number of code chunks required, rounding up for any partial final chunk.
+    # This ensures that all code bytes are covered, even if the code size is not a multiple of CHUNK_SIZE.
+  last_chunk_num = (len(code) + CHUNK_SIZE - 1) // CHUNK_SIZE
 
-When a contract is created, process the write events:
+  for chunk_num in range(last_chunk_num):
+    add_write_access_event(address, (chunk_num + CODE_OFFSET) // 256, (chunk_num + CODE_OFFSET) % 256)
 
-```python
-(
-    address,
-    (CODE_OFFSET + i) // VERKLE_NODE_WIDTH,
-    (CODE_OFFSET + i) % VERKLE_NODE_WIDTH
-)
+  charge_gas()
 ```
 
-For `i` in `0 ... (len(code)+30)//31`.
-
-Note: since no access list existed for code up until this EIP, note that no warm costs are charged for code accesses.
-
 ### Transaction
 
-#### Access events
-
 For a transaction, make these access events:
 
-```
-(tx.origin, 0, BASIC_DATA_LEAF_KEY)
-(tx.origin, 0, CODEHASH_LEAF_KEY)
-(tx.target, 0, BASIC_DATA_LEAF_KEY)
-(tx.target, 0, CODEHASH_LEAF_KEY)
-```
+```python3
+def emit_tx_prologue_events(from: Address, to: Address, value: int):
+  add_write_access_event(from, 0, BASIC_DATA_LEAF_KEY) # Note this is the only write, because of the nonce
+  add_read_access_event(from, 0, CODEHASH_LEAF_KEY)
```

Comment:
Why do we read this key? Is this because of https://eips.ethereum.org/EIPS/eip-3607 ?